src/main/java/nu/studer/gradle/buildscan/teamcity/TeamCityBuildScanPlugin.java

```diff
@@ -22,23 +24,36 @@ public void apply(Project project) {
         if (System.getenv(TEAMCITY_VERSION_ENV) != null && System.getenv(GRADLE_BUILDSCAN_TEAMCITY_PLUGIN_ENV) == null) {
             project.getLogger().quiet(ServiceMessage.of(BUILD_SCAN_SERVICE_MESSAGE_NAME, BUILD_SCAN_SERVICE_STARTED_MESSAGE_ARGUMENT).toString());
 
-            BuildScanExtension buildScanExtension = project.getExtensions().findByType(BuildScanExtension.class);
-            if (supportsScanPublishedListener(buildScanExtension)) {
-                buildScanExtension.buildScanPublished(publishedBuildScan -> {
-                            ServiceMessage serviceMessage = ServiceMessage.of(
-                                    BUILD_SCAN_SERVICE_MESSAGE_NAME,
-                                    BUILD_SCAN_SERVICE_URL_MESSAGE_ARGUMENT_PREFIX + publishedBuildScan.getBuildScanUri().toString()
-                            );
-                            project.getLogger().quiet(serviceMessage.toString());
-                        }
-                );
+            ExtensionContainer extensions = project.getExtensions();
+            try {
+                Class<?> buildScanExtensionClass = Class.forName(BUILD_SCAN_EXTENSION);
+                if (extensions.findByType(buildScanExtensionClass) != null) {
+                    registerCallback(project);
+                    return;
+                }
+            } catch (ClassNotFoundException ignore) {
             }
+            project.getPluginManager().withPlugin(BUILD_SCAN_PLUGIN_ID, appliedPlugin -> registerCallback(project));
```

Comment:
@guylabs @facewindu I don't understand this logic above where we either find the build scan extension or use the plugin manager? Can you please clarify?

Reply:
As the Gradle Enterprise plugin is a settings plugin and is not applied to a `Project` itself, the `pluginManager.withPlugin()` callback would never be called. Therefore we check for the existence of the extension to see if it is already applied via the settings script. If true, we register the callback, if not, we use the  `pluginManager.withPlugin()` callback for the Build Scan plugin, as this is registered on the `Project` level.

Depending when the gradle-build-scan-teamcity-plugin is applied/evaluated one of the above paths will be taken, as the extension class type is the same on both plugins. If we know that the gradle-build-scan-teamcity-plugin is always applied after the Gradle Enterprise or Build Scan plugin, we could get rid of the `pluginManager.withPlugin()` callback, as the extension will always be existent. We kept both to make the application of the gradle-build-scan-teamcity-plugin not order dependent.

Reply:
What is your assumption that the build scan plugin or GE plugin is applied at all? The scenario of the scan/GE plugin not being applied is still not covered, and it's what I have pointed out repeatedly in the tests.

Als just in terms of code style, the current flow is ver hard to follow and needs to be improved.

Reply:
I reworked it a little bit, but the essence stays the same as Guy explained.
I dont find a clever way to achieve all of that.